Profile registration (#1311)
json-api/json-api: extensions/index.md

layout: page
title: Extensions
show_sidebar: true
redirect_from: /profiles
---

JSON:API can be extended with profiles. These profiles enable an API to 
provide clients with information or functionality beyond that described 

> in the base JSON API specification.

Reviewer: ```suggestion
in the base JSON:API specification.
```

Popular profiles may be implemented by off-the-shelf tools so that developers 
can seamlessly take advantage of the features these profiles provide.

## <a href="#existing-profiles" id="existing-profiles" class="headerlink"></a> Existing Profiles

{% for category in site.profile_categories %}
  <h3 id="#profiles-category-{{ category | slugify }}">
    {{ category }}
  </h3>
  <dl class="profiles-list">
    {% for profile in site.profiles %}
      {% if profile.categories contains category %}
        <dt><a href="{% include profile_url.md page=profile %}">{{ profile.name }}</a></dt>
        <dd>{{ profile.short_description }}</dd>
      {% endif %}
    {% endfor %}
  </dl>
{% endfor %}

## <a href="#profile-creation" id="profile-creation" class="headerlink"></a> Creating a New Profile

### <a href="#profile-creation-before" id="profile-creation-before" class="headerlink"></a> Before Creating a Profile

Please **check whether an existing profile fits your needs or could be amended
to fit your needs** before developing a new one.

- If a suitable profile already exists, consider using it. Having fewer, more
widely-deployed profiles makes it easier to create shared tooling.

> - If there's an existing profile that could be amented to fit your needs,

Reviewer 2: ```suggestion
- If there's an existing profile that could be amended to fit your needs,
```

consider asking the profile's author if they would be willing to modify it as
needed. Contact them through the information in their profile's registration
and give them some time to reply.

### <a href="#profile-creation" id="profile-creation" class="headerlink"></a> Authoring Your Profile

To author your profile, [download and fill out the template](/profile_template.md).

### <a href="#profile-register" id="profile-register" class="headerlink"></a> Register & Use Your Profile

> Once it's ready, submit your profile to the JSON:API profile registry by creating

Reviewer 2: ```suggestion
Once your profile is ready, submit it to the JSON:API profile registry by creating
```

Author: Your suggestion has a typo (“submit it profile”); fix that and I’ll gladly pull this in.

Reviewer 2: Whoops! 😅 

Fixed.

a pull request that contains your filled out template as an `index.md` file in a 
uniquely-named folder in the [`_profiles`](https://github.com/json-api/json-api/tree/gh-pages/_profiles) 
directory.
   

> Your submission will be **approved within a week** by one of JSON API's 

Reviewer 2: Really? I think that's an ambitious promise! Are you sure?

Author: Yes. 

My biggest hope for profiles is that they will unblock the community to rapidly experiment with new features, without having to first build consensus for a particular design. The current consensus process has been a huge bottleneck imo.

So, I want these reviews to be quick and not too demanding of authors, which I think means limiting their scope. In particular, my vision for these reviews is that they assess only: 1) is this profile compliant with JSON:API's requirements for profiles?; and 2) if this profile did become widely adopted, would that cause any sort of problems? I don't think the goal should be to assess whether the proposed profile is well-designed or serves a meaningful use case, since that would take too long. If the result is that some poorly-designed profiles get registered, I don't think that's likely to be a big problem.

Reviewer 2: Fair enough!

Reviewer: Personally, I'm unable to make any promise to review profiles within a short timeframe, and I'm not sure it's desirable to force this either. If we or the community see an obvious design flaw in a profile, isn't it preferable to point out the flaw rather than to accept it? (But then again, I'm not comfortable allowing too much extra noise into this issue tracker by using it for extensive reviews.)

All of this to say that I feel that we shouldn't be guaranteeing "approval within a week" let alone "review within a week". I'm more comfortable saying something along the lines of "Your submission will be reviewed in a timely manner by JSON:API's editors and the community."

Author: Yes, if we see a glaring flaw in a profile's design, I definitely think we should point it out. And I imagine that, if we point out a major flaw, the profile's author will generally address it before registering the final version. 

I can clarify the text in this PR to say that, if the community gives feedback, the author can of course take time to act on that feedback, which then would prompt a new round of review, and that this whole process could take more than a week.

I'm really worried, though, about us coming to see the approval process as a "design review", because design reviews tend to take a lot of time/attention, and can get nitpicky. That's frustrating to authors and is likely not a good use of our very-limited bandwidth as editors.

Because profiles are decentralized, I don't see a lot of harm if a poorly-designed profile get registered; people can just not use that profile. It doesn't bind new APIs or hold back the spec. The second review criterion I proposed above ("if this profile did become widely adopted, would that cause any sort of problems?"), which I'm also happy to make more explicit in the PR's text, is meant to ensure that a registered profile wouldn't hold back the ecosystem either.

Given the limited harm, I'd much rather have some bad profiles get through, then have some basically well-designed profiles languish for weeks because we don't have time to review their designs. At least at first, people will likely be developing profiles for their project's immediate use case, and I don't think we can ask them to put their project on hold for more than a week while we're reviewing/asking for changes. 

If the issue is how to help implementers avoid using bad profiles, perhaps we can do something like link to the each profile's review thread next to any links to the its spec. Or, we can exercise our editorial discretion to highlight profiles we think our well designed by listing them above others. Whatever the approach, I don't think this is a hard problem to solve.

All told, I feel very strongly about the importance of quick review, and I'm happy to take on the task of reviewing all incoming profiles according to the criteria I gave above. If the one week timeline proves impossible, we can always change things then. 

Again, I think a critical value of profiles is their potential to unblock the community and get people sharing solutions — which I think means showing the profiles on jsonapi.org — and the last thing I want is for review to become a new bottleneck.

Author: See updated test.

editors, assuming it meets the [profile extension requirements](/format/1.1/#profiles-authoring)
and follows the template above. The community may give you feedback on your
submission before it's approved, but you are not required to act on this 

> feedback.

Reviewer 2: To a casual reader I think this might be understood as: "you are not required to act on feedback for your submission to be approved"

Perhaps something more like this: "the community often provides feedback on new and existing profiles. You are encouraged to accept this feedback, unless it fundamentally alters your profile. This will help the community coalesce around fewer, more powerful profiles."

Author: > I think this might be understood as: "you are not required to act on feedback for your submission to be approved"

That's exactly what I'm trying to convey :) See my comment above.

Reviewer 2: WFM.

One caveat: if your profile defines a new, fundamental mechanism for doing
something "architectural" that other profiles may need to do too, it may be
held for extra consideration. This extra review is designed to check that the
proposed mechanism, were it to become a convention, wouldn't have problematic
ramifications.

Also, note that if the author of a registered profile dies, moves out of contact,

> or otherwise can't or won't make changes that are important to the community,

Reviewer 2: I would simply say "becomes non-responsive for an extended period of time"

Author: I agree that this language could be a bit softer, if that's what you're going for.

For reference, I based it off the language used for media type registrations:

> The IESG may reassign responsibility for a media type. The most common case of this will be to enable changes to be made to types where the author of the registration has died, moved out of contact or is otherwise unable to make changes that are important to the community.

Maybe I should just copy that language directly?

Reviewer 2: I just don't think it's necessary to give specific examples at all. "Legally" it's better to define a class of scenarios which trigger an action than it is to define specific scenarios and risk missing something thereby forcing yourself to interpret "the spirit of the law".

Author: See updated text.
